staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -113,11 +116,22 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 	if err != nil {
 		return apierrors.NewInternalError(err)
 	}
+	if _, ok := attr.VersionedObject.(*unstructured.Unstructured); ok {
+		// Custom Resources don't have corresponding Go struct's.
+		// They are represented as Unstructured.
+		newVersionedObject = &unstructured.Unstructured{}
+	} else {
+		newVersionedObject, err = a.plugin.scheme.New(attr.GetKind())
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+	}
 	// TODO: if we have multiple mutating webhooks, we can remember the json
 	// instead of encoding and decoding for each one.
-	if _, _, err := a.plugin.jsonSerializer.Decode(patchedJS, nil, attr.VersionedObject); err != nil {
+	if _, _, err := a.plugin.jsonSerializer.Decode(patchedJS, nil, newVersionedObject); err != nil {
 		return apierrors.NewInternalError(err)
 	}
+	attr.VersionedObject = newVersionedObject
```

Comment:
can you assign directly to `attr.VersionedObject` above?

Reply:
I actually prefer waiting to overwrite attr.VersionedObject until the decode succeeds

don't we still have the problem that converting attr.VersionedObject back into `attr.Attributes.GetObject()` doesn't necessarily clear the original object?

https://github.com/kubernetes/kubernetes/blob/a7a081b5410fcf711b1f85169225974040051204/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L75-L76

Reply:
> do we still have a problem when ...

You mean "a similar problem in addition", right?

Reply:
> You mean "a similar problem in addition", right?

yes

Reply:
I would leave this `newVersionedObject` as it is too, but my rationale is different. With the brand new empty variable it's clearer that the call `a.plugin.jsonSerializer.Decode()` is supposed to update the variable and not to update the internal state of the called object. For me it took a while to understand where exactly attr.VersionedObject gets updated when I saw the code for the first time.

Reply:
> do we still have a problem when converting attr.VersionedObject back into attr.Attributes.GetObject() if conversion doesn't necessarily clear/reset the original internal object?

This needs a follow-up investigation.